### PR TITLE
Voyages from BigQuery to Apache Beam

### PR DESCRIPTION
**Pull request by smpiano**
- Moved voyages calculation from BigQuery to Apache Beam job.
- the `trip_start` dummy date (`0001-02-03 00:00:00 UTC`) was moved to `null`.
- the `trip_start_anchorage_id` which use to contain `NO_PREVIOUS_DATA` was moved to `null`.
- the `trip_start_visit_id` which use to contain `NO_PREVIOUS_DATA` was moved to `null`.
- the `trip_id` was formatted from `{ssvid}-{vessel_id}-{hex(..)}` to have `{ssvid}-{vessel_id}`. This only applies for init voyages.
- the `trip_end` dummy date was moved to `null`.
- the `trip_end_anchorage_id` which use to contain `ACTIVE_VOYAGE` was moved to `null`.
- the `trip_end_visit_id` which use to contain `ACTIVE_VOYAGE` was moved to `null`.
- the `trip_start` is the **partitioned** field of the output tables.
- the `cluster fields` now are `trip_start, ssvid, vessel_id, trip_id`.

- Adding also the restriction that a voyage should have a duration, so `trip_start` and `trip_end` should be the same.

Screenshots of Graph of the DF job:
Screenshot from 2023-11-07 23-11-19
Screenshot from 2023-11-07 23-10-49

=> Times differences:
**BigQuery**
- BQ voyages_c2: `~1.3h`.
- BQ voyages_c3: `~1.3h`.
- BQ voyages_c4: `~2h`.
Due these queries run in sequential (because they use most of the slots) **Total** `~5h`.
costs: ?

**DF job**
(Using job_id: `2023-11-06_17_47_09-8157413677400264045`)
Time: `33 min 46 sec`.
costs: `$4.15`.

The current  scratch tables results are in : 
- `scratch_matias_ttl_60_days.voyages_duration_c2`
- `scratch_matias_ttl_60_days.voyages_duration_c3`
- `scratch_matias_ttl_60_days.voyages_duration_c4`

### CODE REVIEW

File: pipe_anchorages/transforms/voyages_create.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
+import apache_beam as beam
+import datetime as dt
+import pytz
+
+def create_voyage(visits):
+    # first voyage start from 'we do not know' = None
+    visit_before = {2: None, 3: None, 4: None}
+
+    def min_confidence(v, c:int)->bool:
+        return v['confidence'] >= c
```
Comment (andres-arana) [inline pipe_anchorages/transforms/voyages_create.py:10]: I think this function is too much of an abstraction. I was reading below where it's used and I couldn't figure out what it's supposed to mean. It's clearer to me if I see this:

```python
if (visit and visit["confidence"] >= c and has_duration(visit_before[c], visit)) or (not visit and visit_before[c])):
```

File: pipe_anchorages/transforms/voyages_create.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
+import apache_beam as beam
+import datetime as dt
+import pytz
+
+def create_voyage(visits):
+    # first voyage start from 'we do not know' = None
+    visit_before = {2: None, 3: None, 4: None}
```
Comment (andres-arana) [inline pipe_anchorages/transforms/voyages_create.py:7]: very small comment (feel free to ignore this) but I think `previous_visit` is a better name than `visit_before`.

File: pipe_anchorages/transforms/voyages_create.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        'ssvid': voyage_ssvid,
+        'vessel_id': voyage_vessel_id,
+        'trip_start': voyage_trip_start,
+        'trip_end': visit_after['start_timestamp'] if visit_after else None,
+        'trip_start_anchorage_id': visit_before['end_anchorage_id'] if visit_before else None,
+        'trip_end_anchorage_id': visit_after['start_anchorage_id'] if visit_after else None,
+        'trip_start_visit_id': visit_before['visit_id'] if visit_before else None,
+        'trip_end_visit_id': visit_after['visit_id'] if visit_after else None,
+        'trip_start_confidence': visit_before['confidence'] if visit_before else None,
+        'trip_end_confidence': visit_after['confidence'] if visit_after else None,
```
Comment (andres-arana) [inline pipe_anchorages/transforms/voyages_create.py:36]: I think that there are some abstractions here that can be made. You are doing a bunch of `if dict a is not None, then extract some field out of dict a, else use None`. You could create a function for that like this:

```python
def maybe_get(key, dict, default=None):
    return dict[key] if dict else default
```

In some cases you are doing something like `take key k from either a or b, whichever is not None`, which can be something like this:

```python
def get_from_either(key, first_dict, second_dict):
    return first_dict[key] if first_dict else second_dict[key]
```


and then all of these clauses can be instead something like this:

```python
voyage = {
    'ssvid': get_from_either('ssvid', visit_before, visit_after]),
    'vessel_id': get_from_either('vessel_id', visit_before, visit_after),
    'trip_start': maybe_get('end_timestamp', visit_before),
    # ...
}
```

File: pipe_anchorages/transforms/voyages_create.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    def has_duration(v1, v2):
+        return v1 == None or v1['end_timestamp'] < v2['start_timestamp']
+
+    for visit in visits:
+        for c in [2,3,4]:
+            if (visit and min_confidence(visit, c) and has_duration(visit_before[c],visit)) or (not visit and visit_before[c]):
+                yield build_voyage(c, visit_before[c], visit)
+                visit_before[c] = visit
+
+def build_voyage(trip_confidence, visit_before, visit_after):
```
Comment (andres-arana) [inline pipe_anchorages/transforms/voyages_create.py:21]: I would rename these to `start_visit` and `end_visit` here. You are building a voyage between these 2 visits, so it makes more sense to me.

**Review by andres-arana — COMMENTED** [review]
Very nice work!

One final comment on file structure. In some other pipelines, when you have multiple pipelines per repo I've created a folder that contains everything that belongs to a single pipeline instead of prefixing file names to indicate this. See the segmenter, for example, that contains [this folder](https://github.com/GlobalFishingWatch/pipe-segment/tree/master/pipe_segment/segment_identity). You could have a `voyages` folder on `pipe_anchorages`, which contains a `transforms` folder with the transforms specific to voyages, etc...

**Comment by smpiano** [comment]
Added changes related to the structure, keep everything related to voyages under the `pipe_anchorages/voyages` folder. Removes the abstractions that were not needed, adds the abstractions in the build_voyages method. Replaces the variable names with the suggested in the comments.

**Pull request merged**